lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -177,6 +240,31 @@ def convert_string(value: str) -> str:
     return value
 
 
+def convert_access(value: str) -> str:
+    """
+    Verify that the parameter value is an access scope. Currently this means
+    either "public" or "private".
+
+    NOTE: This is not implemented as an ENUM because it's expected that we'll
+    extend this to support some form of group reference in the future.
+
+    Args:
+        value: parameter value
+
+    Raises:
+        ConversionError: input can't be converted
+
+    Returns:
+        the validated access string
+    """
+    if type(value) is not str:
```

Comment:
Typically in Python one does not use an `if` statement like this, but uses a `try` block to avoid always performing the test and the lowercase operation, when a failed lowercase operation only needs to be handled:

```
try:
    v = value.lower()
except Exception:
    raise ConversionError(value, str.__name__, type(value).__name__)
if v not in Dataset.ACCESS_KEYWORDS:
    raise ConversionError(value, "access", type(value).__name__)
return v
```

Reply:
OK, now this one feels much more unnatural, and I'm not sure it's correct. That is, here we'd be implicitly allowing an instance of *any* class that implements `lower`, which might have a very different semantic than `str.lower()`. In this case, I really prefer to validate that it's a `str` instead of blindly trusting in the existence of a method with a particular name. (I'd *at least* want to be sure that `value.lower()` returned a `str`, but the test I'm making now seems much more straightforward than that test.)

Reply:
Having type checks so deep in the code when they could be made at the outer levels of the interface makes for unnecessary operations.

Seems more unnatural to ever have allowed a different type to make it way all the way down to this level.

Again, why pay the cost of execution for a type check in the lowest levels of the code, when you don't have to?

But if the consensus is that this is the kind of code we want to maintain, it'll have to be that way.

Reply:
I think you're both right.  I think this is a case where the EAFP would be appropriate, but what the code really wants to know is whether the input is a string (the down-casing is a separate thing)...and I don't know of a better way to figure that out than to ask directly.

Reply:
> Having type checks so deep in the code when they could be made at the outer levels of the interface makes for unnecessary operations.

This is a common method to encapsulate a data-driven JSON parameter validation function for the family of query APIs; we'll get here before any "business logic" that depends on types or values, but there's no earlier/shallower place it could be checked without special casing everything in each entry point.